### PR TITLE
feat: add prefer-copy-heredoc rule for file creation patterns

### PR DESCRIPTION
**Pull request by tinovyatkin**
## Summary

- Add new `tally/prefer-copy-heredoc` rule that detects RUN instructions used for file creation and suggests COPY <<EOF syntax
- Extracts file creation from mixed commands (e.g., `apt-get && echo > file && apt-get clean`)
- Supports chmod chaining with `--chmod` flag
- Handles consecutive RUN instructions writing to the same file
- Properly skips append operations (`>>`) to preserve semantics

## Test plan

- [x] Unit tests for file creation detection
- [x] Unit tests for rule logic
- [x] Integration test with real-world Dockerfile
- [x] All existing tests pass
- [x] Linter passes

🤖 Generated with [Claude Code](https://claude.ai/code)

## Summary by CodeRabbit

* **New Features**
  * New lint rule: detects RUN-based file creation and suggests/apply COPY <<EOF heredoc conversions while preserving chmod, mounts, and surrounding commands.

* **Documentation**
  * Rules docs expanded with examples, rationale, options and Supported Rules tally updated to 5.

* **Tests**
  * Extensive new unit and integration tests and updated snapshots; test coverage for many edge cases and fix generation.

* **Other**
  * Example Dockerfile patterns adjusted to prefer heredoc-based file installs.

### CODE REVIEW

File: internal/shell/file_creation.go [hunk 1]
```diff
@@ -0,0 +1,751 @@
...
+		content.WriteString(argContent)
+	}
+
+	result := content.String()
+	if !hasNoNewline {
+		result += "\n"
+	}
+
+	return result, hasUnsafe
+}
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply internal/shell/file_creation.go]: ✅ Addressed in 2822c30. Thanks @coderabbitai\!

File: internal/rules/tally/prefer_copy_heredoc.go [hunk 2]
```diff
@@ -0,0 +1,800 @@
...
+				flushSequence()
+				continue
+			}
+
+			// Update sequence with this file creation
+			sequence, targetPath = updateFileCreationSequence(
+				sequence, targetPath, run, info, flushSequence,
+			)
+			sequenceChmodMode, sequenceChmodRun = "", nil // Clear pending chmod
+			continue
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply internal/rules/tally/prefer_copy_heredoc.go]: ✅ Addressed in cea8e2d. Thanks @coderabbitai[bot]\!

File: internal/shell/file_creation.go [hunk 3]
```diff
@@ -0,0 +1,774 @@
...
+				content:    content,
+				isAppend:   redir.Op == syntax.AppOut,
+			},
+			hasUnsafe: unsafe,
+		}
+	}
+
+	// echo/cat/printf without redirect is not file creation
+	return analyzedCmd{cmdType: cmdTypeOther, text: text}
+}
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply internal/shell/file_creation.go]: ✅ Addressed in ad194bb. Thanks @coderabbitai[bot]\!

File: internal/rules/tally/prefer_copy_heredoc.go [hunk 4]
```diff
@@ -0,0 +1,807 @@
...
+}
+
+// isPathUnder checks if path is under or equal to base directory.
+func isPathUnder(path, base string) bool {
+	// Normalize: ensure base ends with / for proper prefix matching
+	if !strings.HasSuffix(base, "/") {
+		base += "/"
+	}
+	// Check if path equals base (without trailing slash) or is under it
+	return path == strings.TrimSuffix(base, "/") || strings.HasPrefix(path, base)
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply internal/rules/tally/prefer_copy_heredoc.go]: ✅ Addressed in 179bd07. Thanks @coderabbitai[bot]\!

File: internal/rules/tally/prefer_copy_heredoc_test.go [hunk 5]
```diff
@@ -0,0 +1,463 @@
...
+		},
+		{
+			Name: "ssh mount - always safe for literal content",
+			Content: `FROM alpine
+RUN --mount=type=ssh git clone git@github.com:user/repo && echo "cloned" > /app/status
+`,
+			WantViolations: 1, // Safe: SSH doesn't affect file content
+		},
+	})
+}
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply internal/rules/tally/prefer_copy_heredoc_test.go]: ✅ Addressed in 6e0a28c. Thanks @coderabbitai[bot]\!

File: internal/shell/file_creation.go [hunk 6]
```diff
@@ -0,0 +1,813 @@
...
+// extractCatHeredocContentFromStmt finds and extracts heredoc content from a cat statement.
+func extractCatHeredocContentFromStmt(stmt *syntax.Stmt) (string, bool) {
+	// Find the heredoc redirect
+	for _, redir := range stmt.Redirs {
+		if redir.Op == syntax.Hdoc || redir.Op == syntax.DashHdoc {
+			return extractCatHeredocContent(redir)
+		}
+	}
+	// No heredoc found - cat without heredoc creates empty file
+	return "", false
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply internal/shell/file_creation.go]: ✅ Addressed in 0a031fa. Thanks @coderabbitai[bot]\!

File: internal/shell/file_creation.go [hunk 7]
```diff
@@ -0,0 +1,835 @@
...
+
+	return &FileCreationInfo{
+		TargetPath:         targetPath,
+		Content:            content.String(),
+		ChmodMode:          chmodMode,
+		IsAppend:           creations[0].isAppend,
+		HasUnsafeVariables: hasUnsafeVars,
+		PrecedingCommands:  strings.Join(preceding, " && "),
+		RemainingCommands:  strings.Join(remaining, " && "),
+	}
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply internal/shell/file_creation.go]: ✅ Addressed in 39a13e1. Thanks @coderabbitai[bot]\!
